### PR TITLE
Clear Module Settings Cache after deleting settings from database

### LINKED ISSUE #6189: Module settings survive deactivation in cache
### What steps will reproduce the problem?
- Install an optional module that allows settings to be stored, e.g. [HumHub Updater](https://marketplace.humhub.com/module/updater/description)
- Set a setting to a non-default value, e.g. for the updater, set the channel to "beta"
- Deactivate the module
- Re-activate the module
- Check the value you've set (it will be as you set it before, e.g. "beta")
- Clear the cache
- Check the value again (it will now be the default value, e.g. "stable")

### What is the expected result?
- After reactivation, the setting should be again set to it's default value (e.g. "stable" in the above example)
- Deleting the cache should not make any difference

### What do you get instead?
- The setting is as it was before deactivating the module.
- In order to get the default value, you need to clear the cache first

### Additional info

| Q                | A
| ---------------- | ---
| HumHub version   | 1.13.2
| PHP version      | 8.1
| Operating system | Alpine Linux on docker on Ubuntu on WSL

### PR DESCRIPTION
**Pull request by martin-rueegg**
**What kind of change does this PR introduce?**

- Bugfix

**Does this PR introduce a breaking change?** (check one)

- No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the the `master` branch, as it _is_ a hotfix
- [x] Fix #6189
- [ ] All tests are passing
- [ ] New/updated tests are included
- [x] Changelog was modified

**Other information:**
Happy to get feedback and make adjustments, if required.

### CODE REVIEW

**Comment by luke-** [comment]
@martin-rueegg Thanks!

**Comment by martin-rueegg** [comment]
The reasoning for `\humhub\libs\BaseSetting::getCacheKey()` was to have a single source of truth. Rather than creating the key both in the `BaseSetting` as well as in the `\humhub\libs\BaseSettingsManager::getCacheKey()` ...

Also, I tried to do it as flexible as reasonably possible, so that future child classes could still implement the same design pattern.

**Comment by ArchBlood** [comment]
@luke- I would like to note that most of the Setting model is deprecated and we should find a better way for a workaround than using this class, possible solution being migrated to a new class or separating into two classes to keep backwards compatible for v1.1 migration.

**Comment by martin-rueegg** [comment]
> @luke- I would like to note that most of the Setting model is deprecated

Thanks, @ArchBlood. Could you give some indication in what sense the Settings model is deprecated or where to find information on how to do things the right was. As we are in the progress of developing a new module and are currently relating on the Settings model (through the SettingsManager) quite intensely, it would be good to know what to watch out for.

**Comment by ArchBlood** [comment]
You can find which parts of the Setting model that are deprecated by looking for parts that read `@deprecated since version 1.1`

**Comment by ArchBlood** [comment]
As for the `get` method, you can get a module `id` by using `Yii::$app->getModule($moduleId);`, `$moduleId` being the module `id`.

**Comment by martin-rueegg** [comment]
> You can find which parts of the Setting model that are deprecated by looking for parts that read `@deprecated since version 1.1`

Thanks! So basically _every_ method is deprecated, unless it is required for the `ActiveRecord` to work, or it is `Temporary for 1.1. migration`. 🤣 There is only one exception: `Setting::isInstalled()` ...

Hence the correct approach would be to use the `Setting::$value` property, rather than the `Setting::get()` and `Setting::set()` methods, is that correct? As such, the `\humhub\models\Setting` class itself is _not_ deprecated, is that also correct?

If that's the case, I do not see why the proposed changes would mess with the deprecation of the other methods, as
- I _am_ introducing a new base class
- Only overriding a method of the underlying `ActiveRecord`, namely `ActiveRecord::deleteAll()`, 
  plus adding a _new_ static method to get the `getCacheKey()`, because it seemed to be the most logical place to get that information stored/from. :-)

Is there something I am missing here?

However, I notice that the namespace for the new `BaseSetting`, i.e. `humhub\libs` is poorly chosen, and should probably rather be `humhub\models`. Happy to change that, if Lucas sees the implementation to otherwise fit.

No rush though.

File: protected/humhub/libs/BaseSetting.php [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    /**
+     * @param string $moduleId  Name of the module to create the cache key for
+     * @param        ...$values Additional arguments, if required by the static::CACHE_KEY_FORMAT
+     *
+     * @return string The key used for cache operation
+     */
+    public static function getCacheKey(
+        string $moduleId,
+        ...$values
+    ): string {
```
Comment (luke-) [inline protected/humhub/libs/BaseSetting.php:69]:
```suggestion
    public static function getCacheKey(string $moduleId, ...$values): string {
```

File: protected/humhub/libs/BaseSetting.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    /**
+     * @const string Used as the formatting pattern for sprintf when generating the cache key
+     */
+    protected const CACHE_KEY_FORMAT = 'settings-%s';
+
+    public static function deleteAll(
+        $condition = null,
+        $params = []
+    ) {
```
Comment (luke-) [inline protected/humhub/libs/BaseSetting.php:38]:
```suggestion
    public static function deleteAll($condition = null, $params = []) {
```

File: protected/humhub/libs/BaseSetting.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+ * BaseSetting
+ *
+ * @since  1.13.2
+ * @author Martin Rüegg <martin.rueegg@metaworx.ch>
+ */
+abstract class BaseSetting
+    extends ActiveRecord
+{
+
+    // constants
```
Comment (luke-) [inline protected/humhub/libs/BaseSetting.php:24]:
```suggestion
```

File: protected/humhub/libs/BaseSetting.php [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+/**
+ * BaseSetting
+ *
+ * @since  1.13.2
+ * @author Martin Rüegg <martin.rueegg@metaworx.ch>
+ */
+abstract class BaseSetting
+    extends ActiveRecord
+{
```
Comment (luke-) [inline protected/humhub/libs/BaseSetting.php:22]:
```suggestion
abstract class BaseSetting extends ActiveRecord
{
```

File: protected/humhub/libs/BaseSetting.php [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+use Yii;
+use yii\db\ActiveRecord;
+
+/**
+ * BaseSetting
+ *
+ * @since  1.13.2
+ * @author Martin Rüegg <martin.rueegg@metaworx.ch>
+ */
+abstract class BaseSetting
```
Comment (luke-) [inline protected/humhub/libs/BaseSetting.php:20]:
```suggestion
abstract class SettingActiveRecord
```

For base ActiveRecord classes, we currently always add the suffix "ActiveRecord". Normally we also add these in the `components` namespace, but in this case `libs` is also fine.

**Review by luke- — CHANGES_REQUESTED** [review]
@martin-rueegg Thanks, looks fine. A few code style requests...  :-)

**Comment by martin-rueegg** [comment]
> Thank you, looks good! Do you have any idea why the tests fail? Otherwise, I can also take a closer look too.

Nope, seems not have anything to do with the changes, does it?
https://github.com/humhub/humhub/actions/runs/4807191287/jobs/8555632086?pr=6190#step:18:1408

However, I'm currently trying to get my test environment up and running.

**Comment by luke-** [comment]
@martin-rueegg I guess that the failed test has something to do with the PR.

This seems to work like before:
https://github.com/humhub/humhub/blob/master/protected/humhub/modules/user/tests/codeception/acceptance/LoginCest.php#L72-L78
https://github.com/humhub/humhub/blob/master/protected/humhub/modules/user/models/User.php#L990-L994

Maybe some problem with caching?

**Comment by martin-rueegg** [comment]
> @martin-rueegg I guess that the failed test has something to do with the PR.
> 
> This seems to work like before: https://github.com/humhub/humhub/blob/master/protected/humhub/modules/user/tests/codeception/acceptance/LoginCest.php#L72-L78 https://github.com/humhub/humhub/blob/master/protected/humhub/modules/user/models/User.php#L990-L994
> 
> Maybe some problem with caching?

I've been spending most of my afternoon on testing this stuff. But on my local system, with a vanilla database, I do get ton's of errors. None of which has to do with my changes, but many with js.

I find the documentation on testing confusing. And I have no clue how to meaningfully step-debugging the tests.

### What I've done to try to get my test environment running was ...

#### codeception

This did not work for me (https://docs.humhub.org/docs/develop/testing/#test-environment-setup)
``` bash
composer global require "codeception/codeception=2.0.*" "codeception/specify=*" "codeception/verify=*"
```
due to unsolvable dependencies. But the following did:
``` bash
composer global require "codeception/codeception=2.*" "codeception/specify=*" "codeception/verify=*"
```
(Version `2.*` instead of `2.0.*`).

#### DB
``` sql
-- new empty database
CREATE DATABASE `humhub_test` CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;
```

https://community.humhub.com/s/module-development/wiki/Test+Cheatsheet and
https://docs.humhub.org/docs/develop/testing/#test-environment-setup have different instructions regarding db connection the configuration. But I assume it's not really relevant, apart from the later being under version control.

I used the `env.php` and set up `.env`, too. Sourcing it.

Again, the two documentations differ in respect to in which directory to run the migrations, but it does not seem to make a difference, so I did it in the `@webroot`
``` bash
php "$HUMHUB_TEST_YII" migrate/up --includeModuleMigrations=1 --interactive=0
```

Then I seeded the database with:
``` bash
php $PHP_XDEBUG "$HUMHUB_TEST_YII" installer/auto
Install Site:

```
Note the empty result after "Install Site". Is that a problem?

#### Selenium
I used the docker version (from the [cheat sheet](https://community.humhub.com/s/module-development/wiki/Test+Cheatsheet))
``` bash
docker run --detach --net=host --shm-size="2g" selenium/standalone-chrome
```
rather than the java version from [test documentation](https://docs.humhub.org/docs/develop/testing/#run-acceptance-tests)
``` bash
java -Dwebdriver.chrome.driver=chromedriver.exe -jar selenium-server-standalone-2.53.0.jar
```

and the test server:
``` bash
 cd "$HUMHUB_PATH" && /usr/bin/php7.4 --server 127.0.0.1:8080 index-test.php
```

#### Tests
And finally started the tests with
``` bash
source .env

cd "$HUMHUB_PATH" \
 && rm -rf node_modules/ \
 && rm -rf protected/vendor/ \
 && composer install \
 && npm install \
 && git checkout HEAD -- protected/vendor/.gitignore

(rm -rf protected/runtime \
 && git checkout HEAD -- protected/runtime/.gitignore \
 && pushd $HUMHUB_PATH/protected/humhub/tests \
 && /usr/bin/php7.4 $HUMHUB_VENDOR_BIN/codecept run -v)
```

But already the activity tests fail:

[activity]: tests from /home/mdr/projects/humhub/htdocs/protected/humhub/tests/../modules/activity/tests

```
Activity.acceptance Tests (1) ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Modules: WebDriver, \tests\codeception\_support\WebHelper, \tests\codeception\_support\DynamicFixtureHelper
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
E ActivityStreamCest: Test test post creation activity (0.89s)


....


here were 2 errors:

---------
1) ActivityStreamCest: Test test post creation activity
 Test  ../../activity/tests/codeception/acceptance/ActivityStreamCest.php:testSimpleStream

  [Facebook\WebDriver\Exception\JavascriptErrorException] javascript error: $ is not defined
  (Session info: chrome=111.0.5563.146)


Scenario Steps:

 9. $I->executeJS("$('#contentForm_message')[0].click();") at ../../../tests/codeception/_support/AcceptanceTester.php:358
 8. // I am going to create a new post
 7. $I->amOnPage("/space/space?sguid=5396d499-20d6-4233-800b-c6c86e5fa34c") at ../../../tests/codeception/_support/AcceptanceTester.php:123
 6. $I->waitForElementVisible("#wallStream",30) at ../../../tests/codeception/_support/AcceptanceTester.php:233
 5. $I->click("#login-button") at ../../../tests/codeception/_pages/LoginPage.php:27
 4. $I->fillField("Login[password]","123qwe") at ../../../tests/codeception/_pages/LoginPage.php:26

#1  /home/mdr/projects/humhub/htdocs/protected/vendor/php-webdriver/webdriver/lib/Exception/WebDriverException.php:112
#2  /home/mdr/projects/humhub/htdocs/protected/vendor/php-webdriver/webdriver/lib/Remote/HttpCommandExecutor.php:359
#3  /home/mdr/projects/humhub/htdocs/protected/vendor/php-webdriver/webdriver/lib/Remote/RemoteWebDriver.php:605
#4  /home/mdr/projects/humhub/htdocs/protected/vendor/php-webdriver/webdriver/lib/Remote/RemoteWebDriver.php:349
#5  Codeception\Module\WebDriver->executeJS
#6  /home/mdr/projects/humhub/htdocs/protected/humhub/modules/activity/tests/codeception/_support/_generated/AcceptanceTesterActions.php:2723
#7  /home/mdr/projects/humhub/htdocs/protected/humhub/tests/codeception/_support/AcceptanceTester.php:358
#8  /home/mdr/projects/humhub/htdocs/protected/humhub/tests/codeception/_support/AcceptanceTester.php:144
#9  /home/mdr/projects/humhub/htdocs/protected/humhub/modules/activity/tests/codeception/acceptance/ActivityStreamCest.php:28
#10 stream\acceptance\ActivityStreamCest->testSimpleStream
Artifacts:

Png: /home/mdr/projects/humhub/htdocs/protected/humhub/modules/activity/tests/codeception/_output/stream.acceptance.ActivityStreamCest.testSimpleStream.fail.png
Html: /home/mdr/projects/humhub/htdocs/protected/humhub/modules/activity/tests/codeception/_output/stream.acceptance.ActivityStreamCest.testSimpleStream.fail.html

---------

```

Really not sure what to do about this javascript error. Seems like jquery is not loaded?

It also happens with the `master` and the `develop` branches.

Any hint would be appreciated.

**Comment by luke-** [comment]
@martin-rueegg Thats very strange. I can't explain the error message with the missing JQuery.

The global Composer installation might be the problem. Normally I always use the Codeception version which is available in the `protected/vendor/bin` directory and managed via `composer.json`. Meanwhile we also use Codeception 4. I think we should remove this step from the docs.

I just ran the following steps to start a fresh empty HumHub test environment with successful tests:
https://gist.github.com/luke-/fb3c565f630ae7d7c2ab4c51d6ea23aa

**Comment by martin-rueegg** [comment]
Ok, I can now reproduce the issue locally. Will see if I manage to debug it.

**Comment by luke-** [comment]
@martin-rueegg Thanks for the update. Let me know if I can help!

**Review by luke- — APPROVED** [review]

**Comment by martin-rueegg** [comment]
Found and fixed the culprit! :-D

Now, in order to create a test, I need to add data to the contentcontainter_setting table. To do that, a user is required.

- `setting` data is loaded from `protected/humhub/tests/codeception/fixtures/data/setting.php`.
   I thought to use this as an example to create ``setting` data is loaded from `protected/humhub/tests/codeception/fixtures/data/contentcontainer_setting.php`
- `user` data is loaded from `protected/humhub/modules/user/tests/codeception/fixtures/data/user.php`,
   which is in a module.

Hence the following questions:
- Do I need to specify which data is loaded at the beginning of testing?
- Is module data also automatically loaded, or
- can I specify to load the above `user` data or do I need to copy it (with one user only) to `protected/humhub/tests/codeception/fixtures/data/user.php` (and `contentcontainer`)?

**Comment by luke-** [comment]
> Found and fixed the culprit! :-D
> 
> Now, in order to create a test, I need to add data to the contentcontainter_setting table. To do that, a user is required.
> 
> * `setting` data is loaded from `protected/humhub/tests/codeception/fixtures/data/setting.php`.
>   I thought to use this as an example to create ``setting`data is loaded from`protected/humhub/tests/codeception/fixtures/data/contentcontainer_setting.php`
> * `user` data is loaded from `protected/humhub/modules/user/tests/codeception/fixtures/data/user.php`,
>   which is in a module.
> 
> Hence the following questions:
> 
> * Do I need to specify which data is loaded at the beginning of testing?
> * Is module data also automatically loaded, or
> * can I specify to load the above `user` data or do I need to copy it (with one user only) to `protected/humhub/tests/codeception/fixtures/data/user.php` (and `contentcontainer`)?
> * Is it ok to create a unit test, if I want to test just one function, or does it need to be another kind of test, since db data is involved?

Unfortunately, I am not the test expert either. But basically the following standard fixtures should be always loaded: 
https://github.com/humhub/humhub/blob/master/protected/humhub/tests/codeception/_support/DynamicFixtureHelper.php#L121

Per suite or unit test it should be possible to define additional fixtures.

It should be also possible to load fixture data from other modules (e.g. user).  I would prefer to always work with the same defined HumHub test example users, instead of define own per test/suite/module.

A unit test would be sufficient in this case.

I hope I could help a little bit. :-)

**Comment by martin-rueegg** [comment]
Ok, I'm going to make a new PR as there are now tests coming ...

**Comment by luke-** [comment]
@martin-rueegg Sorry, was to fast with merging :-)

**Comment by martin-rueegg** [comment]
Yes, next time I'll add WIP to the title, so you'd know. Unfortunately, I had squashed my chantes locally, so I need to sort them out now ... :-)

**Pull request merged**